packages/wrangler/src/pages/dev.ts

```diff
@@ -462,9 +462,54 @@ export const Handler = async (args: PagesDevArguments) => {
 		logger.debug(`Compiling worker to "${scriptPath}"...`);
 
 		const onEnd = () => scriptReadyResolve();
+		const watchedBundleDependencies: Set<string> = new Set();
+
+		/*
+		 * Pages Functions projects cannot rely on esbuild's watch mode alone.
+		 * That's because the watch mode that's built into esbuild is designed
+		 * specifically for only triggering a rebuild when esbuild's build inputs
+		 * are changed (see https://github.com/evanw/esbuild/issues/3705). With
+		 * Functions, we would actually want to trigger a rebuild every time a new
+		 * file is added to the "/functions" directory.
+		 *
+		 * One solution would be to use an esbuild plugin, and "teach" esbuild to
+		 * watch the "/functions" directory. But it's more complicated than that.
+		 * When we build Functions, one of the steps is to generate the routes
+		 * based on the file tree (see `generateConfigFileFromTree`). These routes
+		 * are then injected into the esbuild entrypoint (see
+		 * `templates/pages-template-worker.ts`). Delegating the "/functions" dir
+		 * watching to an esbuild plugin, would mean delegating the routes generation
+		 * to that plugin as well. This gets very hairy very quickly.
+		 *
+		 * Another solution, is to use a combination of dependencies watching, via
+		 * esbuild, and file system watching, via chokidar. The downside of this
+		 * approach is that a lot of syncing between the two watch modes must be put
+		 * in place, in order to avoid triggering building Functions multiple times
+		 * over one single change (like for example renaming file that's part of the
+		 * dependency graph)
+		 *
+		 * Another solution, which is the one we opted for here, is to delegate file
+		 * watching entirely to a file system watcher, chokidar in this case. While
+		 * not entirely downside-free
+		 *   - we still rely on esbuild to provide us with the dependency graph
+		 *   - we need some logic in place to pre-process and filter the dependencies
+		 *     we pass to chokidar
+		 *   - we need to keep track of which dependencies are being watched
+		 * this solution keeps all things watch mode in one place, makes things easier
+		 * to read, reason about and maintain, separates Pages <-> esbuild concerns
+		 * better, and gives all the flexibility we needed.
+		 */
+		// always watch the "/functions" directory
+		const watcher = watch([functionsDirectory], {
+			persistent: true,
+			ignoreInitial: true,
+		});
+
 		try {
 			const buildFn = async () => {
-				await buildFunctions({
+				const crrBundleDependencies = new Set();
```

Comment:
Can we name this something more verbose? Is `crr` `current`?

Reply:
it is! will do!

Reply:
✅